pkg/cloud/services/compute/instance.go

```diff
@@ -336,6 +369,141 @@ func getPortName(instanceName string, opts *infrav1.PortOpts, netIndex int) stri
 	return fmt.Sprintf("%s-%d", instanceName, netIndex)
 }
 
+func (s *Service) ReconcileRootVolume(eventObject runtime.Object, machine clusterv1.Machine, openStackMachine *infrav1.OpenStackMachine) (bool, *OpenStackError) {
+	rootVolume := openStackMachine.Spec.RootVolume
+
+	// Nothing to do if there's no root volume
+	if rootVolume == nil || rootVolume.Size == 0 {
+		return true, nil
+	}
+
+	resources := &openStackMachine.Status.Resources
+	if resources.RootVolume == nil {
+		resources.RootVolume = &infrav1.OpenStackResourceStatus{}
+	}
+	rootVolumeStatus := resources.RootVolume
+
+	// Nothing to do if the root volume is already created
+	if rootVolumeStatus.Ready {
+		return true, nil
+	}
+
+	volumeID := rootVolumeStatus.UUID
+
+	// Waiting for a previously created root volume to become available
+	if volumeID != "" {
+		createdVolume, err := s.computeService.GetVolume(volumeID)
+		if err != nil {
+			if capoerrors.IsRetryable(err) {
+				return false, transientError(err)
+			}
+			return false, fatalError(err)
+		}
+
+		switch createdVolume.Status {
+		case "available":
+			record.Eventf(eventObject, "VolumeAvailable", "Root volume; id=%s", volumeID)
+			rootVolumeStatus.Ready = true
+			return true, nil
+		case "error":
```

Comment:
this will leave an error volume somewhere
do we delete it somewhere before we create a new one?

Reply:
We won't create a new one. Notice that it returns fatalError(). This means that the controller should put the Machine in a failed state, and the only thing left to do with it is delete the Machine. This is the same as the instance going into an error state.

Because we recorded the volume creation in the status object, when the controller deletes the machine we have a record that we need to cleanup this volume. A 404 here would similarly raise a fatal error because we know to created the volume (we have its uuid), but it no longer exists.

Note that going into the error state is different to, e.g. a 500 returned by cinder because rabbit was unhealthy. In that case we'll return a transientError(), so the controller knows to log it and retry after a back-off.